### DISCUSSION TITLE
support for mutiple db connections

### DISCUSSION THREAD

**Discussion by jzendle** [sqlc-dev/sqlc#908] [Issue Triage]
Hi,

Nice work!

2 quick questions:
I have an application that must connect to multiple databases. Could be mysql and/or PG

1) Does sqlc support multiple connections to different instances of mysql (just mysql)?
2) How about multiple connections to a mix of mysql and PG?

Thanks in advance.

**Comment by kyleconroy** [comment]
> Does sqlc support multiple connections to different instances of mysql (just mysql)?

sqlc will generate a constructor that takes a *sql.DB struct and returns a *Queries struct.

```go
type DBTX interface {
	ExecContext(context.Context, string, ...interface{}) (sql.Result, error)
	PrepareContext(context.Context, string) (*sql.Stmt, error)
	QueryContext(context.Context, string, ...interface{}) (*sql.Rows, error)
	QueryRowContext(context.Context, string, ...interface{}) *sql.Row
}

func New(db DBTX) *Queries {
	return &Queries{db: db}
}
```

You would need to call this constructor with both of your database connections.

> How about multiple connections to a mix of mysql and PG?

sqlc supports this, but it may not work as you expect. sqlc isn't doing any translation of the SQL you right, so you have to make sure that your SQL queries work with MySQL and PostgreSQL, which can be challenging.

**Comment by jzendle** [comment]
Got it. Thanks for the explanation. Sent from my Galaxy
-------- Original message --------From: Kyle Conroy <notifications@github.com> Date: 2/23/21  10:25 PM  (GMT-07:00) To: kyleconroy/sqlc <sqlc@noreply.github.com> Cc: jzendle <joe_zendle@yahoo.com>, Author <author@noreply.github.com> Subject: Re: [kyleconroy/sqlc] support for mutiple db connections (#908) 

Does sqlc support multiple connections to different instances of mysql (just mysql)?

sqlc will generate a constructor that takes a *sql.DB struct and returns a *Queries struct.
type DBTX interface {
	ExecContext(context.Context, string, ...interface{}) (sql.Result, error)
	PrepareContext(context.Context, string) (*sql.Stmt, error)
	QueryContext(context.Context, string, ...interface{}) (*sql.Rows, error)
	QueryRowContext(context.Context, string, ...interface{}) *sql.Row
}

func New(db DBTX) *Queries {
	return &Queries{db: db}
}
You would need to call this constructor with both of your database connections.

How about multiple connections to a mix of mysql and PG?

sqlc supports this, but it may not work as you expect. sqlc isn't doing any translation of the SQL you right, so you have to make sure that your SQL queries work with MySQL and PostgreSQL, which can be challenging.
